File: go/1991.go
package main

func findMiddleIndex(nums []int) int {
	a := 0
	b := 0

	idx := 0

	for i := 1; i < len(nums); i++ {
		b += nums[i]
	}

	if a == b {
		return 0
	}

	for idx < len(nums)-1 {
		a += nums[idx]
		b -= nums[idx+1]

		idx++

		if a == b {
			return idx
		}
	}

	if a == b {
		return len(nums) - 1
	}

	return -1
}

File: zig/1991.zig
const std = @import("std");

fn solution(nums: []const i32) i32 {
    var left: i32 = 0;
    var right: i32 = 0;

    var pivot: usize = 0;

    for (1..nums.len) |i| {
        right += nums[i];
    }

    if (left == right) {
        return @as(i32, @intCast(pivot));
    }

    while (pivot < nums.len - 1) {
        left += nums[pivot];
        right -= nums[pivot + 1];

        pivot += 1;

        if (left == right) {
            return @as(i32, @intCast(pivot));
        }
    }

    if (left == right) {
        return @as(i32, @intCast(nums.len - 1));
    }

    return -1;
}

const expect = std.testing.expectEqual;

test "test case #1" {
    const nums = [_]i32{ 2, 3, -1, 8, 4 };

    const expected = 3;
    const actual = solution(&nums);

    try expect(expected, actual);
}

test "test case #2" {
    const nums = [_]i32{ 1, -1, 4 };

    const expected = 2;
    const actual = solution(&nums);

    try expect(expected, actual);
}

test "test case #3" {
    const nums = [_]i32{ 2, 5 };

    const expected = -1;
    const actual = solution(&nums);

    try expect(expected, actual);
}
